### PR TITLE
Show Streaming errors as View Models

### LINKED ISSUE #252: `Range Error: maximum call stack size exceeded`
### Description

When streaming data using the sdk, the following errors show up in logs
```
error - uncaughtException: RangeError: Maximum call stack size exceeded
    at ListDIDsUseCase.emit (node:events:527:28)
    at ListDIDsUseCase.handleStreamError (webpack-internal:///(api)/./src/lib/core/use-case/list-dids/list-dids-usecase.ts:75:14)
    at ListDIDsUseCase.eval (webpack-internal:///(api)/./src/lib/sdk/usecase.ts:194:59)
    at ListDIDsUseCase.emit (node:events:527:28)
    at ListDIDsUseCase.handleStreamError (webpack-internal:///(api)/./src/lib/core/use-case/list-dids/list-dids-usecase.ts:75:14)
    at ListDIDsUseCase.eval (webpack-internal:///(api)/./src/lib/sdk/usecase.ts:194:59)
    at ListDIDsUseCase.emit (node:events:527:28)
    at ListDIDsUseCase.handleStreamError (webpack-internal:///(api)/./src/lib/core/use-case/list-dids/list-dids-usecase.ts:75:14)
    at ListDIDsUseCase.eval (webpack-internal:///(api)/./src/lib/sdk/usecase.ts:194:59)
    at ListDIDsUseCase.emit (node:events:527:28)
    at ListDIDsUseCase.handleStreamError (webpack-internal:///(api)/./src/lib/core/use-case/list-dids/list-dids-usecase.ts:75:14)
    at ListDIDsUseCase.eval (webpack-internal:///(api)/./src/lib/sdk/usecase.ts:194:59)
    at ListDIDsUseCase.emit (node:events:527:28)
    at ListDIDsUseCase.handleStreamError (webpack-internal:///(api)/./src/lib/core/use-case/list-dids/list-dids-usecase.ts:75:14)
    at ListDIDsUseCase.eval (webpack-internal:///(api)/./src/lib/sdk/usecase.ts:194:59)
    at ListDIDsUseCase.emit (node:events:527:28)
    at ListDIDsUseCase.handleStreamError (webpack-internal:///(api)/./src/lib/core/use-case/list-dids/list-dids-usecase.ts:75:14)
    at ListDIDsUseCase.eval (webpack-internal:///(api)/./src/lib/sdk/usecase.ts:194:59)
    at ListDIDsUseCase.emit (node:events:527:28)
    at ListDIDsUseCase.handleStreamError (webpack-internal:///(api)/./src/lib/core/use-case/list-dids/list-dids-usecase.ts:75:14)
    at ListDIDsUseCase.eval (webpack-internal:///(api)/./src/lib/sdk/usecase.ts:194:59)
    at ListDIDsUseCase.emit (node:events:527:28)
    at ListDIDsUseCase.handleStreamError (webpack-internal:///(api)/./src/lib/core/use-case/list-dids/list-dids-usecase.ts:75:14)
    at ListDIDsUseCase.eval (webpack-internal:///(api)/./src/lib/sdk/usecase.ts:194:59)
    at ListDIDsUseCase.emit (node:events:527:28)
    at ListDIDsUseCase.handleStreamError (webpack-internal:///(api)/./src/lib/core/use-case/list-dids/list-dids-usecase.ts:75:14)
    at ListDIDsUseCase.eval (webpack-internal:///(api)/./src/lib/sdk/usecase.ts:194:59)
    at ListDIDsUseCase.emit (node:events:527:28)
    at ListDIDsUseCase.handleStreamError (webpack-internal:///(api)/./src/lib/core/use-case/list-dids/list-dids-usecase.ts:75:14)
    at ListDIDsUseCase.eval (webpack-internal:///(api)/./src/lib/sdk/usecase.ts:194:59)
    at ListDIDsUseCase.emit (node:events:527:28)
    at ListDIDsUseCase.handleStreamError (webpack-internal:///(api)/./src/lib/core/use-case/list-dids/list-dids-usecase.ts:75:14)
    at ListDIDsUseCase.eval (webpack-internal:///(api)/./src/lib/sdk/usecase.ts:194:59)
    at ListDIDsUseCase.emit (node:events:527:28)
    at ListDIDsUseCase.handleStreamError (webpack-internal:///(api)/./src/lib/core/use-case/list-dids/list-dids-usecase.ts:75:14)
    at ListDIDsUseCase.eval (webpack-internal:///(api)/./src/lib/sdk/usecase.ts:194:59)
    at ListDIDsUseCase.emit (node:events:527:28)
    at ListDIDsUseCase.handleStreamError (webpack-internal:///(api)/./src/lib/core/use-case/list-dids/list-dids-usecase.ts:75:14)
    at ListDIDsUseCase.eval (webpack-internal:///(api)/./src/lib/sdk/usecase.ts:194:59)
    at ListDIDsUseCase.emit (node:events:527:28)
    at ListDIDsUseCase.handleStreamError (webpack-internal:///(api)/./src/lib/core/use-case/list-dids/list-dids-usecase.ts:75:14)
    at ListDIDsUseCase.eval (webpack-internal:///(api)/./src/lib/sdk/usecase.ts:194:59)
    at ListDIDsUseCase.emit (node:events:527:28)
    at ListDIDsUseCase.handleStreamError (webpack-internal:///(api)/./src/lib/core/use-case/list-dids/list-dids-usecase.ts:75:14)
    at ListDIDsUseCase.eval (webpack-internal:///(api)/./src/lib/sdk/usecase.ts:194:59)
    at ListDIDsUseCase.emit (node:events:527:28)
    at ListDIDsUseCase.handleStreamError (webpack-internal:///(api)/./src/lib/core/use-case/list-dids/list-dids-usecase.ts:75:14)
    at ListDIDsUseCase.eval (webpack-internal:///(api)/./src/lib/sdk/usecase.ts:194:59)
    at ListDIDsUseCase.emit (node:events:527:28)
    at ListDIDsUseCase.handleStreamError (webpack-internal:///(api)/./src/lib/core/use-case/list-dids/list-dids-usecase.ts:75:14)
```
This is likely due to missing mechanisms for error handling when streaming leading to spiking in the backpressure somewhere in the chain.

### Steps to reproduce

1. Open listdids page
2. Stream for did data

### Rucio WebUI Version

master

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by maany**
Fix #252

### CODE REVIEW

File: src/lib/infrastructure/presenter/did-meta-presenter.ts
Comment (ThePhisch) [inline src/lib/infrastructure/presenter/did-meta-presenter.ts:1]: Could we not have `status: 'error: ' + errorModel.code`?

**Comment by ThePhisch** [comment]
In the case of streams, the ComDOM will receive a stream of ViewModels in case of success. If an error occurs, only one ViewModel is returned, and it will have `status: 'error'`. My question is: Can multiple errors occur?  If so, will they be combined into one ViewModel or will multiple ViewModels, one for each error, be passed?

**Comment by maany** [comment]
@ThePhisch In case of streams, you can receive
1. One ViewModel with an error, if the backend failed to receive the stream
2. If the backend is able to stream, then you will receive Individual view models with the status as success/error for each element of the stream. 

Elaborating on 2, If there are multiple elements in the stream that have an error, they will lead to 1 error view model per stream element that has an error. For example, consider a stream of response/error models from the usecases `[good, good, bad, good, bad, good]`. Then, at the presenter, it  will convert it into 6 view models, where the status of the view models would look like `[success, success, error, success, error, success]`. 
Hope that helps ;)

File: src/lib/infrastructure/presenter/did-meta-presenter.ts
Comment (maany) [reply src/lib/infrastructure/presenter/did-meta-presenter.ts:1]: Oh, where do you see it? based on the types, status can only be 'success' or 'error'. The error code is not yet embedded into the viewModels.

Comment (maany) [reply src/lib/infrastructure/presenter/did-meta-presenter.ts:1]: The status in a view model can only be 'success' or 'error'. The status constant that you see below is used to send a HTTP response code to the frontend. Not sure where you see status: 'error' + errorModel.code :P

Comment (ThePhisch) [reply src/lib/infrastructure/presenter/did-meta-presenter.ts:1]: no i wanted to suggest adding the HTTP response code to the status stored in the ViewModel. but that might not make sense after all.

**Comment by ThePhisch** [comment]
>     1. One ViewModel with an error, if the backend failed to receive the stream
> 
>     2. If the backend is able to stream, then you will receive Individual view models with the status as success/error for each element of the stream.

What about the case in which the stream only provides one view model which is bad (i.e. `[bad]`)? In this case there is no way to distinguish between an *individual* bad view model in a working stream and a failed stream.

I think it would make sense to explicitly mention that the stream failed entirely, instead of implicitly communicating it (by sending one errorviewmodel and expecting that a stream would always return more than one element, so if the stream has one error and one element, the stream itself must have failed).

(I am currently writing the StreamedTable UI to display the errors, so I noticed this edge case)

**Comment by maany** [comment]
> > ```
> > 1. One ViewModel with an error, if the backend failed to receive the stream
> > 
> > 2. If the backend is able to stream, then you will receive Individual view models with the status as success/error for each element of the stream.
> > ```
> 
> What about the case in which the stream only provides one view model which is bad (i.e. `[bad]`)? In this case there is no way to distinguish between an _individual_ bad view model in a working stream and a failed stream.
> 
> I think it would make sense to explicitly mention that the stream failed entirely, instead of implicitly communicating it (by sending one errorviewmodel and expecting that a stream would always return more than one element, so if the stream has one error and one element, the stream itself must have failed).
> 
> (I am currently writing the StreamedTable UI to display the errors, so I noticed this edge case)

The way to distinguish between whether the failure occurred as a stream element or as a while is to inspect the response code you get back. If the response code was 200 and the view model was in error state, then the stream element encountered an error.

However, if the status code was anything else, then it indicates that the streaming operation failed entirely.

**Pull request merged**